### PR TITLE
feat: no-dts option

### LINKED ISSUE #454: Use typescript without generate d.ts
https://github.com/huozhi/bunchee/blob/fb1214e1bd9d36536189357611909b698ca8b3e5/src/build-config.ts#L212

https://github.com/huozhi/bunchee/blob/fb1214e1bd9d36536189357611909b698ca8b3e5/src/build-config.ts#L258

If I remove "types" in "package.json", should the dts plugin not be applied?

### PR DESCRIPTION
**Pull request by hyoban**
fix #454

### CODE REVIEW

File: src/bin/index.ts [hunk 1]
```diff
@@ -110,13 +113,15 @@ async function run(args: CliArgs) {
     target,
     runtime,
     dts,
+    noDts,
```
Comment (huozhi) [inline src/bin/index.ts:116]: Let's use `dts: boolean`, it should be the same config but with different value

File: src/bin/index.ts [hunk 2]
```diff
@@ -110,13 +113,15 @@ async function run(args: CliArgs) {
...
     env,
     clean,
   } = args
   const cwd = args.cwd || process.cwd()
   const file = args.file ? path.resolve(cwd, args.file) : undefined
   const bundleConfig: BundleConfig = {
     dts,
+    noDts,
```
Comment (huozhi) [inline src/bin/index.ts:124]: merge with `dts`, by default dts is enabled for typescript, if cli dts is set to false, then we disable dts

Comment (hyoban) [reply src/bin/index.ts:124]: But currently, `--dts determine if need to generate types, default: false`

Comment (huozhi) [reply src/bin/index.ts:124]: it's only affecting CLI now, and for ts we're using other condition like hasTsConfig to determine if we should generate types.

when it's off, we should not generate any ts, so we also need to turn off somewhere.

I feel this option wasn't very well designed, it was used for only generating types. but ts project kinda ignore force re-enable it

Comment (hyoban) [reply src/bin/index.ts:124]: So, do we need to change the default option to true. Or, use undefined and explicit false to distinguish.

Comment (huozhi) [reply src/bin/index.ts:124]: I think we can leave the default to `undefined`, which means will still enable in ts. This is only a documentation change.

And we'll also check if dts is `false` when we decide to generate types

**Comment by hyoban** [comment]
Don't know how to use [arg](https://github.com/vercel/arg) to distinguish between `undefined` and `false`. So i keep `no-dts` option in cli, but merge it with `dts`.

**Comment by hyoban** [comment]
> Note that Boolean and [Boolean] have special treatment - an option argument is not consumed or passed, but instead true is returned. These options are called "flags".

It seems we can not let arg to receive `--dts false`

**Comment by devjiwonchoi** [comment]
I see.. I was expecting the behavior of [`--declaration`](https://www.typescriptlang.org/tsconfig/#declaration) in tsc.

IMO expanding the `--no-<arg>` family might not be the best option, but seems reasonable at the moment.

**Comment by hyoban** [comment]
Can i take a look at other cli arg parser?

File: src/bin/index.ts [hunk 3]
```diff
@@ -110,13 +113,14 @@ async function run(args: CliArgs) {
...
     env,
     clean,
   } = args
   const cwd = args.cwd || process.cwd()
   const file = args.file ? path.resolve(cwd, args.file) : undefined
   const bundleConfig: BundleConfig = {
-    dts,
+    dts: dts === true ? true : noDts === true ? false : undefined,
```
Comment (devjiwonchoi) [inline src/bin/index.ts:123]:
```suggestion
    dts: dts || !noDts,
```

**Comment by huozhi** [comment]
We already have the `--no-dts` option value, so we can check it first. I pushed a refactor commit aligning with how we handled others like `--no-external`.

It looks good now! Thanks for the PR!

**Review by huozhi — APPROVED** [review]

**Pull request merged**